kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -78,6 +79,59 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  public static void validateUpdatedSchema(StructType currentSchema, StructType schema) {
+    validateSchema(schema, true);
+    ColumnMapping.validateColumnIds(currentSchema, schema);
+    validateUpdatedSchemaCompatibility(currentSchema, schema);
+  }
+
+  private static void validateUpdatedSchemaCompatibility(
+      StructType currentSchema, StructType newSchema) {
+    // Identify added columns based on field IDs
+    Map<Integer, StructField> newSchemaIdToField = idToField(newSchema);
+    Map<Integer, StructField> currentSchemaIdToField = idToField(currentSchema);
+    for (Map.Entry<Integer, StructField> newFieldEntry : newSchemaIdToField.entrySet()) {
+      if (!currentSchemaIdToField.containsKey(newFieldEntry.getKey())) {
+        if (!newFieldEntry.getValue().isNullable()) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "Cannot add a non-nullable field %s", newFieldEntry.getValue().getName()));
+        }
+      } else {
+        StructField currentField = currentSchemaIdToField.get(newFieldEntry.getKey());
+        StructField newField = newSchemaIdToField.get(newFieldEntry.getKey());
+        if (newField.getDataType() != currentField.getDataType()) {
```

Comment:
Validates no type promotion, this will need to be loosened though as we support those features...

Reply:
This shouldn't use "!=", this should use `equivalent`, otherwise operations like renames fail validation needlessly.